dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
+
+	// if there is no change to be applied then we should keep only
+	// the scheduled changes which belongs to the finalized header
+	// otherwise we should update the scheduled roots to be the child
+	// nodes of the applied scheduled change
+	if changeNode == nil {
+		err := s.keepDescendantScheduledChanges(finalizedHash)
+		if err != nil {
+			return nil, fmt.Errorf("cannot keep descendant scheduled nodes: %w", err)
+		}
+	} else {
+		change = changeNode.change
+		s.scheduledChangeRoots = make([]*pendingChangeNode, len(changeNode.nodes))
+		copy(s.scheduledChangeRoots, changeNode.nodes)
+	}
+
+	return change, nil
+}
+
+// keepDescendantForcedChanges should keep the forced changes for later blocks that
+// are descendant of the finalized block
+func (s *GrandpaState) keepDescendantForcedChanges(finalizedHash common.Hash, finalizedNumber uint) error {
+	onBranchForcedChanges := []*pendingChange{}
+
+	for _, forcedChange := range s.forcedChanges {
+		isDescendant, err := s.blockState.IsDescendantOf(finalizedHash, forcedChange.announcingHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if forcedChange.effectiveNumber() > finalizedNumber && isDescendant {
+			onBranchForcedChanges = append(onBranchForcedChanges, forcedChange)
+		}
+	}
+
+	s.forcedChanges = make(orderedPendingChanges, len(onBranchForcedChanges))
+	copy(s.forcedChanges, onBranchForcedChanges)
+
+	return nil
+}
+
+// keepDescendantScheduledChanges keeps the not applied scheduled changes and remove purged scheduled changes
+func (s *GrandpaState) keepDescendantScheduledChanges(finalizedHash common.Hash) error {
+	onBranchScheduledChanges := []*pendingChangeNode{}
+
+	for _, scheduledChange := range s.scheduledChangeRoots {
+		scheduledChangeHash := scheduledChange.change.announcingHeader.Hash()
+
+		isDescendant, err := s.blockState.IsDescendantOf(finalizedHash, scheduledChangeHash)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchScheduledChanges = append(onBranchScheduledChanges, scheduledChange)
+		}
+	}
+
+	s.scheduledChangeRoots = make([]*pendingChangeNode, len(onBranchScheduledChanges))
+	copy(s.scheduledChangeRoots, onBranchScheduledChanges)
+	return nil
+}
+
+// ApplyScheduledChanges will check the schedules changes in order to find a root
+// that is equals or behind the finalized number and will apply its authority set changes
+func (s *GrandpaState) ApplyScheduledChanges(finalizedHeader *types.Header) error {
+	finalizedHash := finalizedHeader.Hash()
+
+	err := s.keepDescendantForcedChanges(finalizedHash, finalizedHeader.Number)
+	if err != nil {
+		return fmt.Errorf("cannot keep descendant forced changes: %w", err)
+	}
+
+	if len(s.scheduledChangeRoots) == 0 {
+		return nil
+	}
+
+	changeToApply, err := s.getApplicableScheduledChange(finalizedHash, finalizedHeader.Number)
+	if err != nil {
+		return fmt.Errorf("cannot finalize scheduled change: %w", err)
```

Comment:
nit maybe this is more appropriate?

```suggestion
		return fmt.Errorf("cannot get applicable scheduled change: %w", err)
```